crates/biome_js_analyze/src/lint/nursery/no_img_element.rs

```diff
@@ -0,0 +1,137 @@
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```jsx,expect_diagnostic
+    /// <img alt="Foo" />
+    /// ```
+    ///
+    /// ```jsx,expect_diagnostic
+    /// <div>
+    ///   <img alt="Foo" />
+    /// </div>
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```jsx
+    /// <img />
+    /// ```
+    ///
+    /// ```jsx
+    /// <Image src="https://example.com/hero.jpg" />
+    /// ```
+    ///
+    /// ```jsx
+    /// <picture>
+    ///   <source srcSet="https://example.com/hero.avif" type="image/avif" />
+    ///   <source srcSet="https://example.com/hero.webp" type="image/webp" />
+    ///   <img src="https://example.com/hero.jpg" />
+    /// </picture>
+    /// ```
+    ///
+    pub NoImgElement {
+        version: "next",
+        name: "noImgElement",
+        language: "jsx",
+        sources: &[RuleSource::EslintNext("no-img-element")],
+        source_kind: RuleSourceKind::SameLogic,
+        recommended: false,
+    }
+}
+
+declare_node_union! {
+    pub NoImgElementQuery = JsxOpeningElement | JsxSelfClosingElement
+}
+
+impl Rule for NoImgElement {
+    type Query = Ast<NoImgElementQuery>;
+    type State = TextRange;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        node.is_target_valid()?.then(|| node.range())
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, range: &Self::State) -> Option<RuleDiagnostic> {
+        return Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                range,
+                markup! {
+                    "Using "<Emphasis>"<img>"</Emphasis>" could result in slower LCP and higher bandwidth."
+                },
+            )
+            .note(markup! { "Consider using "<Emphasis>"<Image />"</Emphasis>" from "<Emphasis>"next/image"</Emphasis>" to automatically optimize images." })
+            .note(markup! { "This may incur additional usage or cost from your provider." })
+        );
+    }
+}
+
+impl NoImgElementQuery {
+    fn range(&self) -> TextRange {
+        match self {
+            NoImgElementQuery::JsxOpeningElement(jsx) => jsx.range(),
+            NoImgElementQuery::JsxSelfClosingElement(jsx) => jsx.range(),
+        }
+    }
+
+    fn name(&self) -> Option<JsSyntaxToken> {
+        match self {
+            NoImgElementQuery::JsxOpeningElement(jsx) => jsx.name().ok()?.name_value_token(),
+            NoImgElementQuery::JsxSelfClosingElement(jsx) => jsx.name().ok()?.name_value_token(),
+        }
+    }
+
+    fn attributes(&self) -> JsxAttributeList {
+        match self {
+            NoImgElementQuery::JsxOpeningElement(jsx) => jsx.attributes(),
+            NoImgElementQuery::JsxSelfClosingElement(jsx) => jsx.attributes(),
+        }
+    }
+
+    fn get_grandparent(&self) -> Option<JsxOpeningElement> {
+        if let NoImgElementQuery::JsxSelfClosingElement(jsx) = self {
```

Comment:
Nit: this function is only used once, internally, and it doesn't even return the real grandparent, because it does specific casting. The abstraction isn't needed here.

I suggest moving the logic inside `is_target_valid`.